### PR TITLE
Jaliyae/dataloader api

### PR DESCRIPTION
**Pull request by jaliyae**
@goldsborough @apaszke 
This PR includes all the changes we need from the dataloader to support chunk loading. I incorporated most of your suggestions in our previous PRs. This still needs to be integrated with the SharedDataSet change once it is merged. In the  meantime, please have a look and let me know your comments. here is a brief description of what we plan to do with chunk loading.

**Chunk data set** - A dataset that support preloading whole chunks (not a batch) to memory. For this we don't need to know the size of the dataset like in the current data sets. It will be a stateful dataset, so we don't want it to be copied to threads in the dataloader, therefore this PR depends on the SharedDataSet PR. 

**Dataset.reset()** - Similar to dataloader.reset(), this will help us start the preloading and clear Chunk data set at epoch boundaries. 

**Sampler.reset(new_size)** - We need this support, because the size of the dataset or the chunks are unknown at the start, so we need to set the size of the samplers after loading data.

These are the minimum changes we need and as you can see, most of the changes going to be in the chunk data set, which will be completed with all the test cases once we conclude the API changes. 

Please let me know your comments/suggestions.

Previous PRs: [Sampler.rest()](https://github.com/pytorch/pytorch/pull/13870/files), [Dataloader](https://github.com/pytorch/pytorch/pull/13585/files), [SharedDataSet](https://github.com/pytorch/pytorch/pull/13800/files)

### CODE REVIEW

**Comment by jaliyae** [comment]
Ping.

**Comment by ericnakagawa** [comment]
Closing and re-opening to re-run tests.

File: torch/csrc/api/include/torch/data/dataloader.h [hunk 1]
```diff
@@ -279,6 +290,27 @@ std::unique_ptr<DataLoader<Dataset, Sampler>> make_data_loader(
...
+  AT_CHECK(
+      options.chunk_loading() == true,
+      "chunk_loading must be set to true when creating a chunk dataloader.");
+
+  // chunk dataset builds the chunk descriptions in its constructor by
+  // inspecting the whole dataset. It also maintains a set of preload threads.
+  // To avoid copying chunk dataset for each worker, here we wrap it using the
+  // ShareBatchDataset before passing to the dataloader.
+  datasets::SharedBatchDataset<Dataset> shared_chunk_dataset =
+      datasets::make_shared_dataset<Dataset>(dataset);
```
Comment (goldsborough) [inline torch/csrc/api/include/torch/data/dataloader.h]: I'm pretty sure you want to `std::move(dataset)` here

**Review by apaszke — CHANGES_REQUESTED** [review]
Datasets shouldn't have to implement a reset() method. If you really want that, you can always implement that in your own custom dataset class.

As for `Sampler.reset(new_size)`, we've discussed that in the corresponding PR. I do not see why this functionality is necessary, and it really unnecessarily fits the Sampler API to the uses where their only parameters is the dataset size. The fundamental thing you're doing is trying to change the sampler parameters, and that should be done by creating a new sampler.

**Comment by jaliyae** [comment]
@apaszke reset() method allows dataset to perform lazy initialization and recycle its state at epoch boundary. If you look at the proposal, the reset() is a no-op for any dataset that do not want to implement it.

For the sampler reset(), creating a new sampler is fine, but the problem is that Sampler is created by the user and it may have constructor parameters that we don't have access inside the Dataloader or in the Dataset, so creating a sampler is not possible. As we mentioned in the document/meeting, not knowing the size of the dataset is one of the key reasons for this change. We don't know the size of the data until we read it, and hence the two level shuffling.

**Comment by goldsborough** [comment]
@apaszke Just to clarify, I think re-creating the sampler is not an option in C++ because when you have a custom sampler type, say `MySampler`, and you construct it as `MySampler(1, "arg", 3.14)` before passing it into the dataloader, the particular arguments passed to the constructor are lost/no longer available inside the dataloader. So you can't really just construct a fresh one inside for every epoch because you don't know how the user wanted to configure the sampler. That's why I kept the `reset()` method to begin with. However, @jaliyae, I still don't see how you'd use this method in the `DataLoader`. I see no relevant changes in there, only updates to particular samplers. Have you thought further how you'd need to change the dataloader. The thing I'm primarily interested in is how the core DataLoader class needs to change to accommodate your use cases, so that we can evaluate how those change best fit in.

**Comment by jaliyae** [comment]
@goldsborough , Yes, this is exactly the need for the reset() method. I

If you look at the ChunkDataset, we have the two samplers there
  using ChunkSamplerType = ChunkSampler;
  using ExampleSamplerType = ExampleSampler;

Say, we have 100 data files. If we use RandomSampler as the ChunkSampler, then each time a chunk is preloaded (via an async thread) it ask the ChunkSampler for a random chunk index. That is the 100 files are loaded in random order. 
Once a file is loaded, the dataset knows the number of samples in that file, say 10450, and then call ExampleSampler.reset(10450). When the Dataloader request get_bach(32) from the ChunkDataset, it can return 32 examples based on the random indices acquired from the ExampleSampler.

The two samplers live completely within the Dataset, and the Dataloader is created with a BatchSizeSampler which simply returned the provided batch_size in every call. 

Essentially, we have minimum changes to the Dataloader and pushed the preloading to the ChunkDataset.

**Comment by apaszke** [comment]
I understand why you might need reset(), but a good base interface for datasets shouldn't be "everything that someone could possibly need at some point". The reason why I'm trying to keep it simple is exactly that you can easily subclass a dataset yourself, and implement those functions on your own. Then, you can call them at the beginning of the epoch, or at any time you find relevant.

@goldsborough This argument is very unconvincing for me. I see two ways to resolve that problem:
- Ideally we would really use the samplers like we do in Python (so every `begin()` call allocates a new state, and the main sampler object is basically a factory of iterators). That way you really don't need to care about the arguments someone passed in.
- Keep the mutable (bleh) interface as is, and recreate Samplers when you need to change their parameters. I think that's very reasonable since:
  - You really want to change *some* (possibly *all) parameters, so you should provide all of them
  - We avoid hardcoding the assumption that the only Sampler parameter is the dataset size into the API. This is obviously incorrect if you want to do some balancing based on the difficulty of examples, etc.

**Comment by jaliyae** [comment]
@apaszke I agree with the simple interface concept, but please provide a solution as well. We can add any API for the dataset in a subclass, but that won't help unless the dataloader has a way to call it on a generic interface. Could we meet in an online meeting to discuss a solution to read large datasets with the libtorch API?

**Comment by apaszke** [comment]
Yeah we can have a meeting. Why do you have to call those methods through a data loader? I assumed you wanted a SharedDataset, so you can keep references to it from outside the data loader an call whatever methods you want.

I also provided a solution for samplers - create new ones. They’re not expensive objects.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+#include <c10/util/Exception.h>
+#include <torch/data/datasets/base.h>
+#include <torch/data/example.h>
+#include <torch/data/samplers.h>
+
+namespace torch {
+namespace data {
+namespace datasets {
+/// A stateful dataset that support hierarchical sampling and prefetching of
+/// entre chunks. dataset that supports loading an entire chunk of data.
```
Comment (thiagocrepaldi) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: typo

**Comment by ezyang** [comment]
@apaszke @jaliyae What was the resolution of the meeting?

**Comment by jaliyae** [comment]
@ezyang We decided to refactor the dataloader and have a stateless and stateful dataloader implementations. This is the PR that Peter created for this change. It captures all the needs of this PR and we can nicely introduce chunk dataset with this.
https://github.com/pytorch/pytorch/pull/15096

**Comment by ezyang** [comment]
OK, closing!
